### PR TITLE
Improve this page

### PR DESCRIPTION
**Pull request by thehenster**
https://trello.com/c/HN1L3EfK

# Why

The service manual wants a way to allow users to give positive feedback, as well as negative.

# What

A widget to offer a suggestion on how to improve the page or to feedback that the page was useful. This is an alternative to the existing 'Report a Problem' widget that gives encourages positive feedback.

It works by submitting an AJAX request to the feedback frontend, which in turn submits a request to the support-api, which in turn submits the feedback to Zendesk.

Custom event tracking via Google Analytics is supported if the GOV.UK Analytics module is loaded.

The fallback behaviour in the absense of javascript is to send the user to the GOV.UK contact page.

## A widget just above the footer
screen shot 2016-09-16 at 15 55 57

## When you click on "No" or "Is there anything wrong with this page?"
screen shot 2016-09-16 at 15 56 16

## Success
screen shot 2016-09-16 at 15 56 37

# Other moving parts

* The feedback frontend acts as a publicaly accessible proxy inbetween the user's browser and the support-api https://github.com/alphagov/feedback/pull/177
* The support-api receives the request from the feedback frontend and creates a Zendesk ticket https://github.com/alphagov/support-api/pull/96
* gds-api-adapters changed to add helpers and test helpers for the new endpoint https://github.com/alphagov/gds-api-adapters/pull/576

### CODE REVIEW

File: app/assets/javascripts/modules/improve-this-page.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        action: that.pageIsUsefulButton.data('track-action')
+      }
+    }
+
+    this.renderErrors = function (errors) {
+      that.feedbackForm.find('.js-error').remove();
+
+      $.each(errors, function (attrib, messages) {
+        $.each(messages, function (index, message) {
+          var errorNode = $('<div class="improve-this-page__error js-error">' + attrib + ' ' + message + '.</div>');
```
Comment (36degrees) [inline app/assets/javascripts/modules/improve-this-page.js]: I'd be tempted to use `$('<div…>').text(attrib + ' ' + message)` here to escape `attrib` and`message`.

File: app/assets/javascripts/modules/improve-this-page.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+          var field = that.feedbackForm.find('[name="'+ attrib + '"]');
+
+          // If a field with the same name as the error attribute then display
```
Comment (36degrees) [inline app/assets/javascripts/modules/improve-this-page.js]: Think you're missing a word here.

File: app/assets/javascripts/modules/improve-this-page.js [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+      this.controller = new Controller(this.view);
+
+      this.controller.init();
+    }
+  };
+
+  function View ($element) {
+    var that = this;
+
+    this.pageIsUsefulButton = $element.find('.js-page-is-useful');
```
Comment (36degrees) [inline app/assets/javascripts/modules/improve-this-page.js]: Do we prefix the names of jquery objects with `$` here, or did I imagine that? There's a number of places throughout this file where that would apply if so.

File: app/assets/javascripts/modules/improve-this-page.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+      $.each(errors, function (attrib, messages) {
+        $.each(messages, function (index, message) {
+          var errorNode = $('<div class="improve-this-page__error js-error">' + attrib + ' ' + message + '.</div>');
+          var field = that.feedbackForm.find('[name="'+ attrib + '"]');
+
+          // If a field with the same name as the error attribute then display
+          // the error inline with the field. If a matching field doesn't
+          // exist then display it above the form.
+          if (field.length) {
+            field.before(errorNode);
```
Comment (36degrees) [inline app/assets/javascripts/modules/improve-this-page.js]: Is there any thing to clear existing validation messages so that multiple failures don't get duplicated?

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -23,6 +23,45 @@
...
+      </div>
+      <div class="js-feedback-form js-hidden">
+        <div class="grid-row">
+          <div class="column-two-thirds">
+            <div class="js-errors"></div>
+            <form>
+              <input type="hidden" name="path" value="<%= request.original_url -%>">
+              <div class="improve-this-page__section">
+                <label>How should we improve this page?</label>
+                <textarea name="description" rows="5"></textarea>
```
Comment (36degrees) [inline app/views/layouts/application.html.erb]: Worth adding the `required` flag for inline browser validation?

File: spec/javascripts/helpers/with_govuk_analytics.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
+// Temporarily overrides GOVUK.analytics and either returns it
```
Comment (36degrees) [inline spec/javascripts/helpers/with_govuk_analytics.js]: Unfinished comment?

File: app/assets/javascripts/modules/improve-this-page.js [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+  function View ($element) {
+    var that = this;
+
+    this.pageIsUsefulButton = $element.find('.js-page-is-useful');
+    this.offerFeedbackButton = $element.find('.js-offer-feedback');
+    this.feedbackForm = $element.find('.js-feedback-form');
+    this.submitFeedbackButton = that.feedbackForm.find('[type=submit]');
+    this.prompt = $element.find('.js-prompt');
+
+    this.onPageIsUsefulButtonClicked = function (callback) {
```
Comment (36degrees) [inline app/assets/javascripts/modules/improve-this-page.js]: I'm assuming this is how we do things but I find this really hard to follow… when a thing is clicked, click the thing? What?

File: app/assets/javascripts/modules/improve-this-page.js (hunk 3 above)
Comment (thehenster) [inline app/assets/javascripts/modules/improve-this-page.js]: Yeah - we do. I have missed some. I'll update them.

Comment (thehenster) [inline app/assets/javascripts/modules/improve-this-page.js]: There were quite a few in fact! Updated.

File: app/assets/javascripts/modules/improve-this-page.js (hunk 7 above)
Comment (thehenster) [inline app/assets/javascripts/modules/improve-this-page.js]: I've tried to keep the jQuery interacting with the DOM bits separate from the orchestration bit.

It's the idea behind the segregated DOM which I think is cool. http://radar.oreilly.com/2014/01/keeping-jquery-in-check.html http://martinfowler.com/bliki/SegregatedDOM.html

File: app/assets/javascripts/modules/improve-this-page.js (hunk 1 above)
Comment (thehenster) [inline app/assets/javascripts/modules/improve-this-page.js]: Yeah, that's better. Thanks.

File: app/views/shared/_improve_this_page.html.erb [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        <form>
+          <input type="hidden" name="path" value="<%= request.original_url -%>">
+          <div class="improve-this-page__section">
+            <label>How should we improve this page?</label>
+            <textarea name="description" rows="5"></textarea>
+          </div>
+          <p>Include your name and email address if you'd like us to get back to you.</p>
+          <div class="improve-this-page__section">
+            <label>Name</label>
+            <span>(optional)</span>
```
Comment (thehenster) [inline app/views/shared/_improve_this_page.html.erb]: The whole thing is busted for accessibility. I'm hoping to have a chat with Gemma about it.

These labels will need `for` attributes. The yes and no links beside "Is this page useful?" will probably just read out "yes" and "no" and I don't have a good idea how to go about that one. Errors popping up via AJAX sound a little intimidating too, although using the browser validation for the description field will probably be as good as it can be and cover most cases this form is likely to hit.

Redirecting to the contact page, like the JS fallback does, is an option too but I'd much prefer not to.

File: app/views/layouts/application.html.erb (hunk 5 above)
Comment (thehenster) [inline app/views/layouts/application.html.erb]: Good shout, it would be worth adding. I suspect the way I'm submitting the data by plucking out the field values out with `.serialize()` will circumvent the validation too. Hmm.

Comment (thehenster) [inline app/views/layouts/application.html.erb]: I've added commits to bind to the submit event on the form and have added the required attribute. I'm going to leave the code in which renders the errors inline. It's only 3-4 lines and a test which doesn't seem like much of a cost for not having to re-add it if we add other validations to the support-api.

File: app/assets/javascripts/modules/improve-this-page.js [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+        action: that.$pageIsUsefulButton.data('track-action')
+      }
+    }
+
+    this.renderErrors = function (errors) {
+      that.$feedbackFormContainer.find('.js-error').remove();
+
+      $.each(errors, function (attrib, messages) {
+        $.each(messages, function (index, message) {
+          var $errorNode = $('<div/>)', {
```
Comment (36degrees) [inline app/assets/javascripts/modules/improve-this-page.js]: Is that closing parenthesis after `<div/>` meant to be there?

File: app/views/shared/_improve_this_page.html.erb [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+      Is there anything wrong with this page? <span class="visually-hidden"> this page is not useful</span>
+    <% end %>
+  </div>
+  <div class="js-feedback-form js-hidden">
+    <div class="grid-row">
+      <div class="column-two-thirds">
+        <div class="js-errors"></div>
+        <form>
+          <input type="hidden" name="path" value="<%= request.original_url -%>">
+          <input type="hidden" name="user_agent" value="<%= request.user_agent -%>">
```
Comment (36degrees) [inline app/views/shared/_improve_this_page.html.erb]: I need to better understand what ERB escapes when, so forgive me if this is a stupid question – if the user agent or the original url contain quotes or other special characters I'm presuming these'll be escaped as HTML entities?

File: app/views/shared/_improve_this_page.html.erb [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+          <div class="improve-this-page__section">
+            <label for="improve-this-page_description">How should we improve this page? <span class="visually-hidden">required</span></label>
+            <textarea name="description" rows="5" required id="improve-this-page_description"></textarea>
```
Comment (36degrees) [inline app/views/shared/_improve_this_page.html.erb]: To avoid confusion between BEM classes and IDs, is it worth just using a hyphen here (`improve-this-page-description`)? It can be a bit hard to parse the classes and IDs in this block as they're so similar.

(Alternative suggestion – can the association between label and input be made implicit by wrapping the input with the label? Then the `for` and `id` attributes can be removed)

File: app/views/shared/_improve_this_page.html.erb [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+          </div>
+          <p>Include your name and email address if you'd like us to get back to you.</p>
+          <div class="improve-this-page__section">
+            <label  for="improve-this-page_name">Name <span class="improve-this-page__hint">(optional)</span></label>
+            <input type="text" name="name" id="improve-this-page_name">
+          </div>
+          <div class="improve-this-page__section">
+            <label  for="improve-this-page_email">Email <span class="improve-this-page__hint">(optional)</span></label>
```
Comment (36degrees) [inline app/views/shared/_improve_this_page.html.erb]:
```
<label  for="
       ^ 👾 
```

Comment (36degrees) [inline app/views/shared/_improve_this_page.html.erb]: And on :27 too.

File: test/integration/improve_this_page_test.rb [hunk 13]
```diff
@@ -0,0 +1,20 @@
+require 'test_helper'
+
+class ImproveThisPageTest < ActionDispatch::IntegrationTest
```
Comment (36degrees) [inline test/integration/improve_this_page_test.rb]: Any plans to test when javascript is enabled? Even if only that clicking the feedback link makes the relevant form / message visible?

File: app/assets/javascripts/modules/improve-this-page.js (hunk 9 above)
Comment (thehenster) [inline app/assets/javascripts/modules/improve-this-page.js]: I got it from an example in the jQuery docs. Search for "Create some DOM elements" on http://api.jquery.com/jquery/

File: app/views/shared/_improve_this_page.html.erb (hunk 10 above)
Comment (thehenster) [inline app/views/shared/_improve_this_page.html.erb]: Yeah, they should do, unless Rails has made those strings "safe", which I doubt it has. I think it's OK, even desirable, having those values HTML escaped isn't it? If someone has put a script tag in their user agent I quite fancy escaping it.. Ampersands will be escaped in a query string in the URL but we don't have paths based on query strings.. I'm not feeling a lot of urgency to make a change here. Any feelings?

File: app/assets/javascripts/modules/improve-this-page.js (hunk 9 above)
Comment (36degrees) [inline app/assets/javascripts/modules/improve-this-page.js]: Yeah, but that still doesn't have the `)` I'm talking about…

File: app/views/shared/_improve_this_page.html.erb (hunk 10 above)
Comment (36degrees) [inline app/views/shared/_improve_this_page.html.erb]: Yes, I was checking that it _was_ escaped. Not suggesting that it was a problem were that the case.

File: test/integration/improve_this_page_test.rb (hunk 13 above)
Comment (thehenster) [inline test/integration/improve_this_page_test.rb]: I definitely wrote that test once.. I've lost it somehow. I agree it's needed. Ta.

File: app/assets/javascripts/modules/improve-this-page.js (hunk 9 above)
Comment (thehenster) [inline app/assets/javascripts/modules/improve-this-page.js]: Thanks! I saw it in the end :)

File: app/views/shared/_improve_this_page.html.erb (hunk 10 above)
Comment (thehenster) [inline app/views/shared/_improve_this_page.html.erb]: :+1:

File: test/integration/improve_this_page_test.rb [hunk 14]
```diff
@@ -17,4 +17,22 @@ class ImproveThisPageTest < ActionDispatch::IntegrationTest
...
+      within('.improve-this-page') do
+        page.click_link 'No'
+      end
+
+      refute feedback_form_is_hidden
+    end
+  end
+
+  def feedback_form_is_hidden
+    page.find('.js-feedback-form')[:class].include?('js-hidden')
```
Comment (36degrees) [inline test/integration/improve_this_page_test.rb]: I think Capybara already has methods to assert that an element is *actually* visible on the page, rather than tying this to a particular class?

Comment (thehenster) [inline test/integration/improve_this_page_test.rb]: It does.. but they don't work when the CSS that does the hiding isn't available inside the test. I think this is a lesser evil than running a service or "stubbing out" parts of govuk_template.

File: app/views/shared/_improve_this_page.html.erb (hunk 11 above)
Comment (thehenster) [inline app/views/shared/_improve_this_page.html.erb]: I should almost definitely be embarrassed. I didn't know that was valid. It works well with the screenreader too!

File: test/integration/improve_this_page_test.rb (hunk 14 above)
Comment (36degrees) [inline test/integration/improve_this_page_test.rb]: Aaaah of course. OK. Very good. You may continue. Proceed to Go. Collect $200.

Comment (36degrees) [inline test/integration/improve_this_page_test.rb]: Actually, may be worth a comment. ¯\\_(ツ)_/¯

Comment (thehenster) [inline test/integration/improve_this_page_test.rb]: Sure! Added.

File: spec/javascripts/improve-this-page-spec.js [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+          '<a href="/contact/govuk" class="js-offer-feedback">Is there anything wrong with this page?</a>' +
+        '</div>' +
+      '</div>' +
+      '<div class="js-feedback-form js-hidden" data-track-category="improve-this-page" data-track-action="give-feedback">' +
+        '<div class="js-errors"></div>' +
+        '<form>' +
+          '<input type="hidden" name="url" value="http://example.com/path/to/page"></input>' +
+          '<input type="hidden" name="user_agent" value="Safari"></input>' +
+          '<div>' +
+            '<label>How should we improve this page?</label>' +
```
Comment (gemmaleigh) [inline spec/javascripts/improve-this-page-spec.js]: The markup in the fixture doesn't match the markup you are using. 
You've got labels wrapping the label text, hint text and input field - but not here.

Comment (thehenster) [inline spec/javascripts/improve-this-page-spec.js]: Thanks for this - I've added a commit.

In the past I've managed to get javascript tests loading the rails partial as a fixture which avoids this kind of problem. I'll have a play in a separate branch.
